File: Algorithm-CPP/42_이분검색.cpp
#include <iostream>
using namespace std;

void swap(int arr[], int idx1, int idx2) {
    int temp = arr[idx1];
    arr[idx1] = arr[idx2];
    arr[idx2] = temp;
}

int Parition(int arr[], int left, int right) {
    int pivot = arr[left];
    int low = left + 1;
    int high = right;

    while(low <= high) {
        while(pivot > arr[low]) low++;
        while(pivot < arr[high]) high--;

        if(low <= high) swap(arr, low, high);
    }

    swap(arr, left, high);
    return high;
}

void quicksort(int arr[], int left, int right) {
    if(left <= right) {
        int pivot = Parition(arr, left, right);
        quicksort(arr, left, pivot - 1);
        quicksort(arr, pivot + 1, right);
    }
}

int main() {
    int n, m;
    cin >> n >> m;

    int arr[n];
    for(int i = 0; i < n; i++) cin >> arr[i];

    quicksort(arr, 0, n);

    for(int i = 0; i < n; i++) {
        if(arr[i] == m) {
            cout << i + 1 << endl;
            break;
        }
    }

    return 0;
}

File: Algorithm-Swift/Algorithm-Swift/Inflearn/42_이분검색.swift
//
//  42_이분검색.swift
//  Algorithm-Swift
//
//  Created by SHIN YOON AH on 2023/08/18.
//

import Foundation

func Partition(arr: inout [Int], left: Int, right: Int) -> Int {
    let pivot = arr[left]
    var low = left + 1
    var high = right

    while low <= high {
        while arr[low] < pivot { low += 1 }
        while arr[high] > pivot { high -= 1 }

        if low <= high { arr.swapAt(low, high) }
    }

    arr.swapAt(left, high)

    return high
}

func quicksort(_ arr: inout [Int], left: Int, right: Int) {
    if left <= right {
        let pivot = Partition(arr: &arr, left: left, right: right)
        quicksort(&arr, left: left, right: pivot - 1)
        quicksort(&arr, left: pivot + 1, right: right)
    }
}

func solution42() {
    let nm = readLine()!.split(separator: " ").map { Int($0)! }
    let n = nm.first!
    let m = nm.last!
    var arr = readLine()!.split(separator: " ").map { Int($0)! }

    quicksort(&arr, left: 0, right: n - 1)

    arr.enumerated().forEach { index, num in
        if num == m {
            print(index + 1)
            return
        }
    }
}
